lesson_17/sources/dynamic_int_array.cpp

```diff
@@ -0,0 +1,105 @@
+#include <iostream>
+#include <algorithm>
+
+#include "../headers/dynamic_int_array.hpp"
+
+
+//default c-tor
+dynamic_int_array::dynamic_int_array() : dynamic_int_array{ 0 }
+{
+}
+
+//c-tor with size
+dynamic_int_array::dynamic_int_array(const std::size_t size) : size_{ 0 }
+{
+    max_size_ = size > initial_buffer_size ? size : initial_buffer_size;
+    dynamic_buffer_ = new int[max_size_];
+}
+
+//copy c-tor
+dynamic_int_array::dynamic_int_array(const dynamic_int_array& other) :
+    size_{ other.size_ }, max_size_{ other.max_size_ }
+{
+    dynamic_buffer_ = new int[max_size_];
+    std::copy_n(other.dynamic_buffer_, size_, dynamic_buffer_);
+}
+
+//free allocated memory
+dynamic_int_array::~dynamic_int_array()
+{
+    delete[] dynamic_buffer_;
+    dynamic_buffer_ = nullptr;
+}
+
+// copy assignment operator
+dynamic_int_array& dynamic_int_array::operator=(const dynamic_int_array& other)
+{
+    if (this != &other)
+    {
+        delete[] dynamic_buffer_;
+        dynamic_buffer_ = nullptr;
+
+        max_size_ = other.max_size_;
+        size_ = other.size_;
+
+        dynamic_buffer_ = new int[max_size_];
+        std::copy_n(other.dynamic_buffer_, size_, dynamic_buffer_);
+    }
+
+    return *this;
+}
+
+int& dynamic_int_array::operator[](const std::size_t index)
+{
+    if (index >= size_)
+    {
+        std::cerr << "dynamic_int_array: error. operator[] index: " << index << " is out of range!";
+    }
+
+    return dynamic_buffer_[index];
+}
+
+//delete previous memory, create new inner array with updated size
+void dynamic_int_array::set_size(const std::size_t new_size)
+{
+    delete[] dynamic_buffer_;
+    dynamic_buffer_ = nullptr;
+
+    size_ = 0;
+    max_size_ = new_size > initial_buffer_size ? new_size : initial_buffer_size;
+    dynamic_buffer_ = new int[max_size_];
+}
```

Comment:
You should store data that was in array. If new size is less then previous crop the array, if it is bigger add zeros

Reply:
done